### PR TITLE
Nc news 2

### PR DESCRIPTION
**Pull request by MJahidH**

### CODE REVIEW

File: __tests__/test-data.test.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+const db = require("../db/connection");
+const seed = require("../db/seeds/seed");
+const data = require("../db/data/test-data/index");
+
+beforeEach(() => {
+  return seed(data);
+});
+
+afterAll(() => {
+  db.end();
```
Comment (ghost) [inline __tests__/test-data.test.js]: this needs a return statement as it won't end the connection asynchronously.

File: __tests__/test-data.test.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+});
+
+describe.only("api/topics", () => {
```
Comment (ghost) [inline __tests__/test-data.test.js]: Don't send work through with .only in the test file. Please remove and resubmit.

**Review by ghost — CHANGES_REQUESTED** [review]
Don't leave .only in your test files. Please remove this and address the afterall statement and resubmit.

File: __tests__/test-data.test.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+beforeEach(() => {
+  return seed(data);
+});
+
+afterAll(() => {
+  db.end();
+});
+
+describe("api/topics", () => {
+  test("recieve status code 200", () => {
```
Comment (ghost) [inline __tests__/test-data.test.js]: What are we getting a status code of 200 for? I know its for topics, so do you, but could this be more descriptive?

File: __tests__/test-data.test.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    return request(app).get("/api/topics").expect(200);
+  });
+  test("returns an array ", () => {
```
Comment (ghost) [inline __tests__/test-data.test.js:17]: Could this not have been combined into one test? We can chain expecting a 200 status code and then if it is an array containing the objects we are expecting.

File: __tests__/test-data.test.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    return request(app)
+      .get("/api/topics")
+      .expect(200)
+      .then((body) => {
+        const arrayOfTopics = body._body;
```
Comment (ghost) [inline __tests__/test-data.test.js]: Why is this on a key of _body when that is a private key.

File: __tests__/test-data.test.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        expect(Array.isArray(arrayOfTopics)).toBe(true);
+      });
+  });
+  test("check if each obj in array has the keys slug and description  ", () => {
+    return request(app)
+      .get("/api/topics")
+      .expect(200)
+      .then((body) => {
+        const arrayOfTopics = body._body;
+        const areKeysPresent = arrayOfTopics.every((obj) => {
```
Comment (ghost) [inline __tests__/test-data.test.js:33]: You can use for each to write an expect statement for each object. But this does sort of work as well.

File: __tests__/test-data.test.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        expect(areKeysPresent).toBe(true);
+      });
+  });
+
+  test("returns 404 error if topics is not present after /api ", () => {
+    return (
+      request(app)
+        .get("/api/topic")
+        // wrong end point name
+        .expect(404)
```
Comment (ghost) [inline __tests__/test-data.test.js:49]: Would there be a nice error message to go with this?

File: app.js [hunk 8]
```diff
@@ -0,0 +1,12 @@
+const express = require('express');
+const {getTopics} = require("./controllers/topics.controllers") 
+const app = express();
+
+app.use(express.json());
```
Comment (ghost) [inline app.js]: You aren't posting data yet, this is not needed.

File: models /topics.model.js [hunk 9]
```diff
@@ -0,0 +1,6 @@
+const db = require("../db/connection")
+
+
+exports.requireTopics = () => {
+    return db.query("SELECT * FROM topics")
```
Comment (ghost) [inline models /topics.model.js]: Do you want to return the whole response or just the rows?

File: package.json [hunk 10]
```diff
@@ -28,7 +28,9 @@
   },
   "dependencies": {
     "dotenv": "^16.0.0",
-    "pg": "^8.7.3"
+    "express": "^4.18.2",
+    "pg": "^8.11.3",
+    "supertest": "^6.3.3"
```
Comment (ghost) [inline package.json:33]: Is this a dependency or a dev dependency? Will we need it when hosting to production?

File: controllers/topics.controllers.js [hunk 11]
```diff
@@ -0,0 +1,7 @@
+const { requireTopics } = require("../models /topics.model");
+
+exports.getTopics = (req, res) => {
+  requireTopics().then((data) => {
+    res.status(200).send(data.rows);
```
Comment (ghost) [inline controllers/topics.controllers.js]: This should be an object with a key of topics, not just the array. Trust me, you'll thank us in front end.

**Review by ghost — CHANGES_REQUESTED** [review]
Don't be afraid to combine tests. This is integration testing not unit testing, you don't need a test for the status code, then the data type, then the data value. That can all be asserted in one. There really only is two tests for this ticket.

I'm a little concerned by accessing data on the _body key. If you've sent back the rows you should be able to see this on the body. Don't forget that body is a key on the response, so just because you've used body in your then block does not mean you've accessed it automatically.

I'd like you to address the comments in this pull request before you move on. Some minor changes to make, but almost there.

File: controllers/topics.controllers.js [hunk 12]
```diff
@@ -1,7 +1,8 @@
 const { requireTopics } = require("../models /topics.model");
 
 exports.getTopics = (req, res) => {
-  requireTopics().then((data) => {
-    res.status(200).send(data.rows);
+  requireTopics().then((topics) => {
+  
+    res.status(200).send({topics});
```
Comment (nc-mick-fay) [inline controllers/topics.controllers.js:6]: This is a great change to send your topics back on a descriptive key

File: __tests__/test-data.test.js [hunk 13]
```diff
@@ -42,7 +41,7 @@ describe("api/topics", () => {
       });
   });
 
-  test("returns 404 error if topics is not present after /api ", () => {
+  test("200 , returns 404 error if topics is not present after /api ", () => {
```
Comment (nc-mick-fay) [inline __tests__/test-data.test.js:44]: The original test description was more clear, as now I'm not sure if you expect a 200 or 404.

Add an error message after your expect statement as well just so we know what your user can expect back

Comment (nc-mick-fay) [reply __tests__/test-data.test.js:44]: This test is actually testing express's native 404 response for unhandled endpoints. You don't need to test this as it comes as standard with the package

Comment (nc-mick-fay) [reply __tests__/test-data.test.js:44]: If you were to create a custom 404 error handler, then a test like this would be needed

**Review by nc-mick-fay — APPROVED** [review]
All your changes look good and ready to merge into your main branch.

Your 404 error handler is a bit unnecessary as you don't currently have a custom 404 error handler - this is testing express's native 404 response for unhandled endpoints

**Pull request merged**